### PR TITLE
RxFire Realtime Database

### PR DESCRIPTION
**Pull request by davideast**
WIP on API

### CODE REVIEW

File: packages/rxfire/database/fromRef.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import { Observable } from 'rxjs';
+import { map, delay, share } from 'rxjs/operators';
+import { ListenEvent, SnapshotPrevKey } from './interfaces';
+
+/**
+ * Create an observable from a Database Reference or Database Query.
+ * @param ref Database Reference
+ * @param event Listen event type ('value', 'added', 'changed', 'removed', 'moved')
+ */
+export function fromRef(
```
Comment (jshcrowthe) [inline packages/rxfire/database/fromRef.ts]: This method allows users to dynamically invoke various methods off of a reference. It's kind of scary and I would rather us support only the `on` and have users handle the `once` case themselves.

File: packages/rxfire/test/database.test.ts [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+   * functions are useful if the process becomes brittle or tedious.
+   * Note that removing is less necessary since the tests are run
+   * offline.
+   *
+   * Note: Database tests do not run exactly the same offline as
+   * they do online. Querying can act differently, tests must
+   * account for this.
+   */
+  beforeEach(() => {
+    app = initializeApp({
```
Comment (jshcrowthe) [inline packages/rxfire/test/database.test.ts]: Test using the user specified test config please.

File: packages/rxfire/database/fromRef.ts (hunk 1 above)
Comment (davideast) [reply packages/rxfire/database/fromRef.ts]: fixed

File: packages/rxfire/database/list/audit-trail.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { database } from 'firebase';
+import { Observable } from 'rxjs';
+import { SnapshotPrevKey, ChildEvent } from '../interfaces';
+import { fromRef } from '../fromRef';
+import { map, withLatestFrom, scan, skipWhile } from 'rxjs/operators';
+import { stateChanges } from './';
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/database/list/audit-trail.ts]: fixed

File: packages/rxfire/database/interfaces.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+import { database } from 'firebase';
+
+export type QueryFn = (ref: database.Reference) => database.Query;
+export type ChildEvent =
+  | 'child_added'
+  | 'child_removed'
+  | 'child_changed'
+  | 'child_moved';
+export type ListenEvent = 'value' | ChildEvent;
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/database/interfaces.ts]: fixed

File: packages/rxfire/database/interfaces.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { database } from 'firebase';
+
+export type QueryFn = (ref: database.Reference) => database.Query;
+export type ChildEvent =
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/database/interfaces.ts]: fixed

File: packages/rxfire/test/database.test.ts (hunk 2 above)
Comment (davideast) [reply packages/rxfire/test/database.test.ts]: Not sure what you mean seeing how this is an automated test.

**Comment by mikelehen** [comment]
FYI- I have not looked at this. Are you looking for / waiting for my review?  If so, please add me to the assignees and let me know if there are any specific things you want feedback on.  Else if you just need Josh's review for packaging, that's great too (I'm under water with reviews at the moment so my review would likely be delayed).

Also, can you update the PR title and description if this is past the "WIP" stage?

**Comment by Kamshak** [comment]
Wohoo this is 🔥🔥🔥
I suppose angularfire will be updated later to simply use this API?

**Comment by davideast** [comment]
@mikelehen Sorry, Michael! Github auto-asked for your review. Nothing is needed on your end. I'll take everyone but Josh off the list from now on.

**Comment by mikelehen** [comment]
@davideast No worries at all. I just wanted to make sure nothing was waiting on me.

File: packages/rxfire/test/database.test.ts (hunk 2 above)
Comment (jshcrowthe) [reply packages/rxfire/test/database.test.ts]: The config here points to a static firebase project. As part of the test setup for the repo we have users provide a config. I want us to reference that config here if we are actually going to hit prod for our tests

Comment (davideast) [reply packages/rxfire/test/database.test.ts]: fixed

**Pull request merged**